beam/src/Rhyolite/Task/Beam/Worker.hs

```diff
@@ -0,0 +1,151 @@
+    -- 3. Perform the serializable task
+    withTransactionSerializableRunBeamPostgres dbConn $ do
+      primaryKeyAndInput <- runSelectReturningOne $ select $
+        -- Only lock one row (limit_ 1)
+        lockingAllTablesFor_ PgSelectLockingStrengthUpdate Nothing $ limit_ 1 $ do
+          task <- all_ table
+
+          -- Both task fields should be empty for an unclaimed task
+          -- Also apply any other filters that may have been passed, using ready
+          guard_ $ isNothing_ (task ^. field . taskResult) &&. isNothing_ (task ^. field . taskCheckedOutBy) &&. ready
+
+          -- Return the primary key (task id) along with a custom field that the user asked for.
+          pure (primaryKey task, task ^. field . taskPayload)
+
+      -- In case we did not find any rows, no update SQL will be run
+      -- The row lock that we acquired above will be reset when the transaction ends.
+      forM primaryKeyAndInput $ \(taskId, input) -> do
+        -- Mark the retrieved task as checked out, by the current worker
+        runUpdate $
+          update table
+            (\task -> (task ^. field . taskCheckedOutBy) <-. val_ (Just workerName))
+            (\task -> primaryKey task ==. val_ taskId)
+
+        runSerializableInsideTransaction dbConn (Logger logger) $ (,) taskId <$> go input
+    -- END Transaction, release row lock
+
+  case checkedOutValue of
+    Nothing -> pure False
+    Just (taskId, action) -> do
+      -- Get the followup Serializable
+      followup <- action
+
+      -- BEGIN Transaction
+      withTransactionSerializableRunBeamPostgres dbConn $ do
+        -- Get the result value from the serializable
+        b <- runSerializableInsideTransaction dbConn (Logger logger) followup
+
+        -- Update the task's result field, set checked out field to null
+        runUpdate $ update table
+          (\task -> mconcat
+            [ task ^. field . taskResult <-. val_ (Just b)
+            , task ^. field . taskCheckedOutBy <-. val_ Nothing])
+          (\task -> primaryKey task ==. val_ taskId)
+      -- END Transaction
+
+      pure True
+
+-- | Run a worker thread
+-- The worker will wake up whenever the timer expires or the wakeup action is called
+-- Once woken up, the worker will be run repeatedly until it reports that it was not able to find any work to do; then it will start sleeping for the given duration.
+-- If the wakeup action is called while the worker is running, the worker will run again as soon as it finishes, even if it returns False.  This is necessary because otherwise, since checking for work isn't generally atomic, there would be a race condition: worker starts (e.g. enters looking-for-work transaction), work is created, wakeup called, worker finishes with no work found, sleep.
+withWorker :: (MonadIO m, MonadBaseControl IO m) => NominalDiffTime -> IO Bool -> (IO () -> m a) -> m a
+withWorker d work child = do
```

Comment:
Should we be using `supervise` here? Or should the caller of this function decide whether it ought to be supervised?

Reply:
I think `withWorker` already keeps searching for tasks, once awake. What benefit would `supervise` add?

Reply:
`supervise` handles restarting on exceptions thrown

Reply:
That's a good question... supervise just logs the exceptions to the terminal, and I can imagine someone wanting to do something slightly different, but also we usually would want to make sure that the taskWorker thread doesn't somewhat quietly disappear on us.

Reply:
Ok I'm resolving this because the groundhog version doesn't supervise and it looks like from the original implementation there was an intention that the worker thread can be reaped in certain cases.